File: 401-500/0494-Find-Target-Sub-Ways/0494.py
class Solution:
  def findTargetSubWays(self, nums: List[int], target: int) -> int:
    s = sum(nums)
    if (target + s) & 1: return 0

    V = (target + s) >> 1
    f = [1] + [0] * V
    for n in nums:
      for i in range(V, n - 1, -1):
        f[i] += f[i - n]
    return f[-1]

File: 401-500/0494-Find-Target-Sub-Ways/0494.java
class Solution {
  public int findTargetSubWays(int[] nums, int target) {
    int sum = 0;
    for (int num : nums) {
      sum += num;
    }
    int diff = sum - target;
    if (diff < 0 || diff % 2 != 0) {
      return 0;
    }
    int neg = diff / 2;
    int[] dp = new int[neg + 1];
    dp[0] = 1;
    for (int num : nums) {
      for (int j = neg; j >= num; j--) {
        dp[j] += dp[j - num];
      }
    }
    return dp[neg];
  }
}